### ISSUE TITLE
Vector panic when attempting to process metrics like logs in VRL

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#12467]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

With the below VRL, Vector panics when the input is metrics rather than logs.

### Configuration

```text
data_dir = "/var/lib/vector/"

[sources.source0]
namespace = "host"
scrape_interval_secs = 2
type = "host_metrics"

[transforms.transform0]
inputs = ["source0"]
timezone = "local"
type = "remap"
source = """
# Parse the received .ddtags field so we can more easily access the contained tags, set to empty object if parsing fails
.ddtags = parse_key_value(.ddtags, key_value_delimiter: ":", field_delimiter: ",") ?? {}
.ddtags.sender = "vector"
# Re-encode Datadog tags as a string for the `datadog_logs` sink
.ddtags = encode_key_value(.ddtags, key_value_delimiter: ":", field_delimiter: ",")
"""

[sinks.sink0]
inputs = ["transform0"]
type = "console"
encoding.codec = "json"
```

### Version

vector 0.21.1

### Debug Output

```text
2022-04-28T21:46:36.468720Z  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info,buffers=info,kube=info"
2022-04-28T21:46:36.468793Z  INFO vector::app: Loading configs. paths=["/tmp/tmp.toml"]
2022-04-28T21:46:36.471245Z  INFO vector::topology::running: Running healthchecks.
2022-04-28T21:46:36.471515Z  INFO vector: Vector has started. debug="false" version="0.21.1" arch="aarch64" build_id="none"
2022-04-28T21:46:36.471602Z  INFO vector::app: API is disabled, enable by setting `api.enabled` to `true` and use commands like `vector top`.
2022-04-28T21:46:36.471796Z  INFO vector::topology::builder: Healthcheck: Passed.
2022-04-28T21:46:36.482197Z ERROR transform{component_kind="transform" component_id=transform0 component_type=remap component_name=transform0}: vector::internal_events::remap: Mapping failed with event. error="function call error for \"encode_key_value\" at (311:384): expected object, got null" error_type="conversion_failed" stage="processing" internal_log_rate_secs=10
thread 'vector-worker' panicked at 'event will be set', src/transforms/remap.rs:429:48
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'vector-worker' panicked at 'internal error: entered unreachable code: join error or bad poll', src/topology/builder.rs:655:30
2022-04-28T21:46:36.482811Z ERROR transform{component_kind="transform" component_id=transform0 component_type=remap component_name=transform0}: vector::topology: An error occurred that Vector couldn't handle.
2022-04-28T21:46:36.486726Z  INFO vector: Vector has stopped.
2022-04-28T21:46:36.487443Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="source0, sink0" time_remaining="59 seconds left"
```

**Comment by JeanMertz** [comment]
So, this is what appears to be happening:

1. `encode_key_value` fails at runtime, even though it shouldn't be allowed to (there's no `!`, abort-on-error)
2. because there's no `!`, `can_fail()` returns `false`
3. which in turn means the remap transform impl expects `event` to be cloned as expected
4. which then triggers the `unwrap` and the panic

Given that this only happens for metrics, and the fact that the function-call happens in the `.ddtags = ...` assignment expression, _and_ the fact that I know that we handle assignments for metrics differently, my first assumption is that this happens somewhere in that part of the code.

Essentially, for logs, this code would not compile requesting the author to use infallible assignment (`.ddtags, err = ...`), but apparently that's not happening in this case, and I wonder why.

**Comment by JeanMertz** [comment]
So looking at the error:

> function call error for \"encode_key_value\" at (311:384): expected object, got null

`encode_key_value` requires an object as its first argument, which in the case of this program:

```coffee
.ddtags = parse_key_value(.ddtags, key_value_delimiter: ":", field_delimiter: ",") ?? {}
.ddtags.sender = "vector"
.ddtags = encode_key_value(.ddtags, key_value_delimiter: ":", field_delimiter: ",")
```

Is `.ddtags`.

What is happening here, is that the compiler marks the type definition of `ddtags` as an object, which is accurate, given the second line of that program is `.ddtags.sender`. But then at runtime, it appears `.ddtags` is actually `null`.

Again, this must have something to do with the custom assignment logic we have for metrics, I'm still digging why this can happen.

**Comment by JeanMertz** [comment]
Yeah, so the problem is that this is an invalid program for metrics.

Metric tags need to be assigned to `.tags`, e.g. `.tags.sender = "vector"`, not `.ddtags`.

In this case, the following error is triggered:  https://github.com/vectordotdev/vector/blob/50990404cb1f5ef6b555f0946c7f4a862071ca71/lib/vector-core/src/event/vrl_target.rs#L122-L128

However, we've made assignments infallible (assuming the rhs expression is infallible): https://github.com/vectordotdev/vector/blob/50990404cb1f5ef6b555f0946c7f4a862071ca71/lib/vrl/compiler/src/expression/assignment.rs#L285-L287

Otherwise, we would've had to add special syntax to handle this case, but in the case of logs, inserting a value can never fail, so we didn't have a need for this. With metrics, however, we _do_ allow insertion to fail, and that failure is silently ignored.

This is what then makes the compiler assume the program is infallible, while at runtime the program will actually fail because the assignment actually never happens, and thus the query for `.ddtags` returns `null`.

**Comment by jblang** [comment]
I hit a similar issue:

```
2022-06-29T22:40:56.447497Z ERROR transform{component_kind="transform" component_id=parse_gc component_type=remap component_name=parse_gc}: vector::internal_events::remap: Mapping failed with event. error="function call error for \"compact\" at (113:161): function call error for \"split\" at (121:160): function call error for \"strip_whitespace\" at (127:152): expected string, got null" error_type="conversion_failed" stage="processing" internal_log_rate_secs=10
thread 'vector-worker' panicked at 'event will be set', src/transforms/remap.rs:459:48
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'vector-worker' panicked at 'internal error: entered unreachable code: join error or bad poll', src/topology/builder.rs:669:30
2022-06-29T22:40:56.697647Z ERROR transform{component_kind="transform" component_id=parse_gc component_type=remap component_name=parse_gc}: vector::topology: An error occurred that Vector couldn't handle.
2022-06-29T22:40:56.727766Z  INFO vector: Vector has stopped.
2022-06-29T22:40:56.774686Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="output_caches, output_tombstone_warnings, output_tombstone_errors, splunk, unnest_threadpools, unnest_generations, output_aggregation_warnings, output_tpc_local_backpressure, output_gc_pause, output_gc_generation, output_hints_stored, route_statuslogger, unnest_caches, output_unlogged_batches, output_large_partitions, output_threadpools, unnest_memtables, output_network_backpressure, input_events, file_input, output_bufferpools, split_statuslogger, statuslogger_format, route, filtered_events, coalesce_old_statuslogger, output_node_flaps, output_hints_delivered, output_memtables, output_large_batches, output_dropped_messages, output_stolen_tasks, output_local_pauses, output_fingerprint_exception, parsed_events, parse_gc, final_events, output_tpc_core_backpressure" time_remaining="59 seconds left"
thread 'vector-worker' panicked at 'unit error: ()', lib/vector-core/src/fanout.rs:369:42
2022-06-29T22:40:56.796326Z ERROR transform{component_kind="transform" component_id=route component_type=route component_name=route}: vector::topology: An error occurred that Vector couldn't handle.
thread 'vector-worker' panicked at 'unit error: ()', lib/vector-core/src/fanout.rs:369:42
2022-06-29T22:40:56.805135Z ERROR transform{component_kind="transform" component_id=filtered_events component_type=filter component_name=filtered_events}: vector::topology: An error occurred that Vector couldn't handle.
thread 'vector-worker' panicked at 'unit error: ()', lib/vector-core/src/fanout.rs:369:42
2022-06-29T22:40:56.860925Z ERROR transform{component_kind="transform" component_id=parsed_events component_type=remap component_name=parsed_events}: vector::topology: An error occurred that Vector couldn't handle.
thread 'vector-worker' panicked at 'unit error: ()', lib/vector-core/src/fanout.rs:369:42
2022-06-29T22:40:56.873721Z ERROR transform{component_kind="transform" component_id=input_events component_type=remap component_name=input_events}: vector::topology: An error occurred that Vector couldn't handle.
thread 'vector-worker' panicked at 'unit error: ()', lib/vector-core/src/fanout.rs:369:42
thread 'vector-worker' panicked at 'join error: JoinError::Panic(Id(24), ...)', src/topology/builder.rs:212:30
2022-06-29T22:40:56.883711Z ERROR source{component_kind="source" component_id=file_input component_type=file component_name=file_input}: vector::topology: An error occurred that Vector couldn't handle.
```

The transform in question is:

```
  parse_gc:
    type: remap
    inputs:
      - route.gc_pause
    source: |-
      temp = split(.message, r' GC in |ms\. ') ?? []
      .gc_type = temp[0]
      .duration = to_int(temp[1]) ?? 0
      .sizes = compact(split(strip_whitespace(temp[2]), r'; '))
```

This transform parses messages like this:

```
G1 Old Generation GC in 6757ms.  Compressed Class Space: 7658856 -> 7657720; G1 Eden Space: 7935623168 -> 0; G1 Old Gen: 10723988672 -> 3175154624; G1 Survivor Space: 318767104 -> 0; Metaspace: 70479288 -> 70470600
```

I have not yet been able to identify the specific message that is causing the crash.

**Comment by spencergilbert** [comment]
Duplicate of https://github.com/vectordotdev/vector/issues/17217, to be resolved by https://github.com/vectordotdev/vrl/issues/186
